File: src/main/java/com/hazza/leetcode/SegmentTwo/No96UniqueBinarySearchTrees.java
package com.hazza.leetcode.SegmentTwo;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/unique-binary-search-trees/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-10-11
 * Time: 10:19 AM
 */
public class No96UniqueBinarySearchTrees {
    public int numTrees(int n) {
        int[] dp = new int[n + 1];

        dp[0] = dp[1] = 1;
        for (int i = 2; i <=n ; i++)
            for (int j = 1; j <= i; j++)
                dp[i] += dp[j - 1] * dp[i - j];

        return dp[n];
    }
}

File: src/main/python/leetcode_by_python/tree/No96_Unique_Binary_Search_Trees.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-08-22
# https://leetcode.com/problems/unique-binary-search-trees/


class Solution:
    def numTrees(self, n: int) -> int:
        dp = [0] * (n + 1)

        dp[0] = dp[1] = 1
        for i in range(2, n + 1):
            for j in range(1, i + 1):
                dp[i] += dp[j - 1] * dp[i - j]

        return dp[n]

